spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+before `FinalizeBlock`, even if one of them might end up corresponding to the
+decided block and thus have to be reexecuted upon `FinalizeBlock`.
+
+### States and ABCI++ Connections
+
+#### Consensus Connection
+
+The Consensus Connection should maintain an `ExecuteTxState` - the working state
+for block execution. It should be updated by the call to `FinalizeBlock`
+during block execution and committed to disk as the "latest
+committed state" during `Commit`. Execution of a proposed block (via `PrepareProposal`/`ProcessProposal`) **must not** update the `ExecuteTxState`, but rather be kept as a separate candidate state until `FinalizeBlock`
+confirms which of the candidate states (if any) can be used to update `ExecuteTxState`.
+
+#### Mempool Connection
+
+The mempool Connection should maintain a `CheckTxState`
+to sequentially process pending transactions in the mempool that have
+not yet been committed. It should be initialized to the latest committed state
+at the end of every `Commit`.
+
+During the execution of a Consensus instance, the `CheckTxState` may be updated concurrently with the
+`ExecuteTxState`, as messages may be sent concurrently on the Consensus and Mempool connections.
+At the end of the Consensus instance, as described above, Tendermint locks the mempool and flushes
+the mempool connection before calling `Commit`. This ensures that all pending `CheckTx` calls are
+responded to and no new ones can begin.
+
+After `Commit`, while still holding the mempool lock, `CheckTx` is run again on all transactions
+that remain in the node's local mempool after filtering those included in the block.
+An additional `Type` parameter is made available to the `CheckTx` function that
+indicates whether an incoming transaction is new (`CheckTxType_New`), or a
+recheck (`CheckTxType_Recheck`).
+
+Finally, after re-checking transactions in the mempool, Tendermint will unlock
+the mempool connection. New transactions are once again able to be processed through `CheckTx`.
+
+Note that `CheckTx` is just a weak filter to keep invalid transactions out of the mempool.
+Since the transaction cannot be guaranteed to be checked against the exact same state as it
+will be executed as part of a (potential) decided block, `CheckTx` shouldn't check *everything*
+that affects the transaction's validity, in particular those checks whose validity may depend on
+transaction ordering. `CheckTx` is weak because a Byzantine node need not care about `CheckTx`;
+it can propose a block full of invalid transactions if it wants. The mechanism ABCI++ has
+in place for dealing with such behavior is `ProcessProposal`.
+
+##### Replay Protection
+
+To prevent old transactions from being replayed, `CheckTx` must implement
+replay protection.
+
+It is possible for old transactions to be sent to the application. So
+it is important `CheckTx` implements some logic to handle them.
```

Comment:
I don't know that this is necessarily true.

Reply:
This is taken from the original text, but it seems reasonable to me...

Reply:
I think the current TM data store implementation is difficult to figure out if the transaction is an old transaction or not unless tx indexer is enabled and the mempool is able to query it (by txhash?). The App should implement logic to be able to reject the transaction so let it cannot be stored in the mempool.

Reply:
Yes, as we are talking about `CheckTx` here, we _imply_ that it is in principle up to the Application to implement replay protection.
I'll clarify this in the text

Reply:
Revamped the whole section